src/compiler/crystal/codegen/debug.cr

```diff
@@ -154,33 +317,49 @@ module Crystal
       end
     end
 
-    private def declare_local(type, alloca, location)
+    private def declare_local(type, alloca, location, basic_block : LLVM::BasicBlock? = nil)
       location = location.try &.original_location
-      return unless location
+      return false unless location
+
+      file, dir = file_and_dir(location.filename)
+      @current_debug_file = file = @debug_files[location.filename] ||= di_builder.create_file(file, dir)
 
       debug_type = get_debug_type(type)
-      return unless debug_type
+      return false unless debug_type
 
       scope = get_current_debug_scope(location)
-      return unless scope
-
-      file, dir = file_and_dir(location.filename)
-      file = di_builder.create_file(file, dir)
+      return false unless scope
 
       var = yield scope, file, location.line_number, debug_type
       expr = di_builder.create_expression(nil, 0)
-
-      di_builder.insert_declare_at_end(alloca, var, expr, builder.current_debug_location, alloca_block)
+      if basic_block
+        block = basic_block
+      else
+        basic_blocks = context.fun.basic_blocks.to_a
```

Comment:
Instead of allocating an entire array of the basic blocks, which can be many... what do you think if we use `each` and we just keep the last one? If it's `nil` it means there are none, otherwise we have the last one.

Reply:
Yeah, we can do it this way.